src/sensai/vector_model.py

```diff
@@ -164,41 +251,50 @@ def predict(self, x: pd.DataFrame) -> pd.DataFrame:
         """
         x = self._computeInputs(x)
         y = self._predict(x)
+        # TODO or not TODO: why do we do this? Isn't it the implementation's responsibility to get it right?
         y.index = x.index
         y = self._outputTransformerChain.apply(y)
         if self._targetTransformer is not None:
             y = self._targetTransformer.applyInverse(y)
         return y
 
-    @abstractmethod
-    def _predict(self, x: pd.DataFrame) -> pd.DataFrame:
-        pass
-
-    def fit(self, X: pd.DataFrame, Y: pd.DataFrame):
+    def fit(self, X: pd.DataFrame, Y: pd.DataFrame = None):
         """
         Fits the model using the given data
 
         :param X: a data frame containing input data
-        :param Y: a data frame containing output data
+        :param Y: a data frame containing output data or None for unsupervised models
         """
         log.info(f"Training {self.__class__.__name__}")
-        self._predictedVariableNames = list(Y.columns)
-        X = self._computeInputs(X, y=Y)
-        if self._targetTransformer is not None:
-            self._targetTransformer.fit(Y)
-            Y = self._targetTransformer.apply(Y)
+        X = self._computeInputs(X, y=Y, fit=True)
+        if Y is None and not isinstance(self._targetTransformer, _IdentityDFT):
+            raise ValueError(
+                f"targetTransformer should be trivial be since no target data is provided. "
+                f"Instead got: {self._targetTransformer.__class__.__name__}"
+            )
+        self._targetTransformer.fit(Y)
+        Y = self._targetTransformer.apply(Y)
+
         self._modelInputVariableNames = list(X.columns)
-        self._modelOutputVariableNames = list(Y.columns)
-        log.info(f"Training with outputs[{len(self._modelOutputVariableNames)}]={self._modelOutputVariableNames}, inputs[{len(self._modelInputVariableNames)}]=[{', '.join([n + '/' + X[n].dtype.name for n in self._modelInputVariableNames])}]")
+        output_info = ""
+        if Y is not None:
+            self._predictedVariableNames = list(Y.columns)
```

Comment:
This is broken. Please never change the order of statements without due cause. predictedVariableNames and modelOutputVariableNames are not the same thing. For an explanation, see the docstring of getModelOutputVariableNames.

Reply:
Nice catch, I did not want that to happen. Will go through the code again (once we settled on how to deal with unsupervised models)

Reply:
Done by reverting to previous version of the code